File: src/litelm/llm.py
# Copyright 2026 Emmanuel Decitre
# SPDX-License-Identifier: Apache-2.0

from .bridge.runtime import ensure_runtime


class LLM:
    DEFAULTS = {
        "generator_model": "gpt2",
        "embedding_model": "all-MiniLM-L6-v2",
        "max_new_tokens": 50,
        "temperature": 0.7,
        "top_k": 50,
        "do_sample": True,
        "use_local_models": False,
        "local_models_path": "/drive/models",
        "auto_detect_local": True,  # Automatically use local models if available
    }

    def __init__(self, js_instance):
        self.js = js_instance

    # --- lifecycle ---

    @classmethod
    async def create(cls, **overrides):
        import os

        ensure_runtime()

        from js import createLLM

        config = {**cls.DEFAULTS, **overrides}

        # Auto-detect local models if enabled
        if config.get("auto_detect_local", True) and not config.get("use_local_models", False):
            generator_model = config.get("generator_model", cls.DEFAULTS["generator_model"])
            local_model_path = f"{config['local_models_path']}/Xenova/{generator_model}"
            zip_path = f"{config['local_models_path']}/xenova-{generator_model}.zip"

            # Check for zip file first, then unzip if needed
            if os.path.exists(zip_path) and not os.path.exists(local_model_path):
                print(f"Found zipped model at {zip_path}, extracting...")
                import zipfile

                with zipfile.ZipFile(zip_path, "r") as zipf:
                    zipf.extractall(config["local_models_path"])
                print(f"Extracted to {config['local_models_path']}")

            if os.path.exists(local_model_path):
                print(f"Found local model at {local_model_path}, using local files")
                config["use_local_models"] = True
                # Convert filesystem path to emfs: URI for browser access
                # emfs: expects path relative to /drive, so models/Xenova/gpt2
                # Remove /drive/ prefix to get models/Xenova/gpt2
                emfs_path = local_model_path.replace("/drive/", "")
                config["emfs_model_uri"] = f"emfs:{emfs_path}"
                print(f"Using emfs URI: {config['emfs_model_uri']}")
            else:
                print(f"No local model found at {local_model_path}, will download from HuggingFace")

        instance = createLLM(config)
        await instance.init()

        return cls(instance)

    # --- core API ---

    async def generate(self, prompt: str):
        return await self.js.generate(prompt)

    async def embed(self, text: str):
        vec = await self.js.embed(text)
        return list(vec)

    # --- similarity ---

    async def similarity_search(self, query, docs):
        from js import cosineSimilarityBatch

        query_vec = await self.embed(query)

        doc_vecs = []
        for doc in docs:
            doc_vecs.append(await self.embed(doc))

        sims = cosineSimilarityBatch(query_vec, doc_vecs)

        results = list(zip(docs, sims))
        results.sort(key=lambda x: x[1], reverse=True)

        return results

    # --- model export ---

    async def export_model_files(self, model_name=None, as_zip=True):
        """Export model files from browser cache to /drive for download.

        Args:
            model_name: Name of the model to export (default: current generator model)
            as_zip: If True, export as a zip file; if False, export as individual files
        """
        import os
        import zipfile

        if model_name is None:
            model_name = self.DEFAULTS["generator_model"]

        # Check if model is already exported
        if as_zip:
            zip_path = f"/drive/models/xenova-{model_name}.zip"
            if os.path.exists(zip_path):
                message = f"Model already exported at {zip_path}"
                print(message)
                return message
        else:
            model_dir = f"/drive/models/Xenova/{model_name}"
            if os.path.exists(model_dir) and os.listdir(model_dir):
                message = f"Model already exported at {model_dir}"
                print(message)
                return message

        result = await self.js.exportModelFiles(model_name)

        # Check if result is an object with files (Python-side export needed)
        if hasattr(result, "success") and result.success:
            print(f"Exporting {len(result.files)} files for {result.model}")

            if as_zip:
                # Export as zip file
                zip_path = f"/drive/models/xenova-{result.model}.zip"
                os.makedirs("/drive/models", exist_ok=True)

                # Deduplicate files by name, keeping the largest version
                files_by_name = {}
                for file_obj in result.files:
                    filename = file_obj.filename
                    data = file_obj.data
                    file_bytes = bytes(data.to_py())

                    if filename not in files_by_name or len(file_bytes) > len(files_by_name[filename]):
                        files_by_name[filename] = file_bytes

                with zipfile.ZipFile(zip_path, "w", zipfile.ZIP_DEFLATED) as zipf:
                    for filename, file_bytes in files_by_name.items():
                        # Store with Xenova/{model}/ prefix
                        arcname = f"Xenova/{result.model}/{filename}"
                        zipf.writestr(arcname, file_bytes)
                        print(f"Added to zip: {arcname} ({len(file_bytes)} bytes)")

                message = f"Successfully exported {len(files_by_name)} files to {zip_path}"
                print(message)
                return message
            else:
                # Export as individual files
                model_dir = f"/drive/models/Xenova/{result.model}"
                os.makedirs(model_dir, exist_ok=True)
                print(f"Created directory: {model_dir}")

                # Deduplicate files by name, keeping the largest version
                files_by_name = {}
                for file_obj in result.files:
                    filename = file_obj.filename
                    data = file_obj.data
                    file_bytes = bytes(data.to_py())

                    if filename not in files_by_name or len(file_bytes) > len(files_by_name[filename]):
                        files_by_name[filename] = file_bytes

                for filename, file_bytes in files_by_name.items():
                    filepath = f"{model_dir}/{filename}"
                    with open(filepath, "wb") as f:
                        f.write(file_bytes)

                    print(f"Exported: {filepath} ({len(file_bytes)} bytes)")

                message = f"Successfully exported {len(files_by_name)} files to {model_dir}"
                print(message)
                return message
        else:
            # String result from JS side
            print(f"Export result: {result}")
            return result

File: src/lmlite/llm.js
// lmLite JS runtime

class BrowserLLM {
  constructor(config) {
    this.generatorModel = config.generator_model;
    this.embeddingModel = config.embedding_model;
    this.useLocalModels = config.use_local_models ?? false;
    this.localModelsPath = config.local_models_path ?? '/drive/models';
    this.emfsModelUri = config.emfs_model_uri ?? null;

    this.max_new_tokens = config.max_new_tokens ?? 50;
    this.temperature = config.temperature ?? 0.7;
    this.top_k = config.top_k ?? 50;
    this.do_sample = config.do_sample ?? true;

    this.generator = null;
    this.embedder = null;
  }

  async init() {
    const { pipeline, env } = await import(
      "https://cdn.jsdelivr.net/npm/@xenova/transformers"
    );

    // Configure for browser environment
    if (this.useLocalModels) {
      env.allowLocalModels = false;
      env.useBrowserCache = false;
      env.allowRemoteModels = true; // Keep true but intercept fetch

      // Set up custom backend to read from Pyodide FS
      const pyodide = globalThis.pyodide;
      if (pyodide) {
        env.backends.onnx.wasm.proxy = false; // Disable web worker for local files

        // Custom fetch that reads from Pyodide FS
        const originalFetch = globalThis.fetch;
        globalThis.fetch = async (url, options) => {
          const urlStr = typeof url === 'string' ? url : url.toString();

          // Check if this is a model file request for HuggingFace
          if (urlStr.includes('huggingface.co') &&
              (urlStr.includes(`Xenova/${this.generatorModel}`) ||
               (this.embeddingModel && urlStr.includes(`Xenova/${this.embeddingModel}`)))) {

            // Extract filename from URL
            const parts = urlStr.split('/');
            const filename = parts[parts.length - 1];
            const modelName = parts[parts.indexOf('Xenova') + 1];
            const filepath = `${this.localModelsPath}/Xenova/${modelName}/${filename}`;

            console.log(`Intercepting HF request, reading from Pyodide FS: ${filepath}`);

            try {
              const data = pyodide.FS.readFile(filepath);
              const blob = new Blob([data]);

              return new Response(blob, {
                status: 200,
                statusText: 'OK',
                headers: new Headers({
                  'Content-Type': 'application/octet-stream',
                  'Content-Length': data.length.toString()
                })
              });
            } catch (err) {
              console.error(`Failed to read ${filepath}:`, err);
              return new Response(null, { status: 404, statusText: 'Not Found' });
            }
          }

          // Fall back to original fetch for non-model requests
          return originalFetch(url, options);
        };

        console.log(`Using local models from Pyodide FS: ${this.localModelsPath}`);
      }
    } else {
      env.allowLocalModels = false;
      env.useBrowserCache = true;
      console.log('Using remote models with browser cache (IndexedDB)');
    }

    try {
      let modelPath;
      if (this.useLocalModels) {
        // Use HuggingFace path format - our custom fetch will intercept
        modelPath = `Xenova/${this.generatorModel}`;
      } else {
        // Download from HuggingFace
        modelPath = `Xenova/${this.generatorModel}`;
      }

      console.log(`Loading text-generation model: ${modelPath}`);

      // Add progress callback
      this.generator = await pipeline(
        "text-generation",
        modelPath,
        {
          // Only use quantized models when downloading from HuggingFace
          // Local models are usually not quantized
          quantized: !this.useLocalModels,
          // Explicitly use WASM backend
          device: 'wasm',
          // Progress callback
          progress_callback: (progress) => {
            if (progress.status === 'downloading') {
              console.log(`Downloading ${progress.file}: ${progress.progress?.toFixed(2)}%`);
            } else if (progress.status === 'done') {
              console.log(`Downloaded ${progress.file}`);
            } else if (progress.status === 'ready') {
              console.log(`Model ready: ${progress.file}`);
            } else if (progress.status === 'progress') {
              console.log(`Loading ${progress.file}: ${progress.progress?.toFixed(2)}%`);
            }
          }
        }
      );

      console.log(`Model loaded successfully: ${this.generatorModel}`);

      // Warm up the model with a dummy inference to ensure it's fully ready
      console.log(`Warming up model...`);
      await this.generator("test", { max_new_tokens: 1 });
      console.log(`Model warm-up complete`);

    } catch (error) {
      throw new Error(
        `Failed to load text-generation model '${this.generatorModel}': ${error.message}. ` +
        `This may be due to model incompatibility with browser WASM/WebGPU. ` +
        `Try a different model like 'gpt2' or check model availability.`
      );
    }

    if (this.embeddingModel) {
      try {
        // Always use HuggingFace format - custom fetch intercepts if local
        const embeddingPath = `Xenova/${this.embeddingModel}`;

        console.log(`Loading embedding model: ${embeddingPath}`);
        this.embedder = await pipeline(
          "feature-extraction",
          embeddingPath
        );
        console.log(`Embedding model loaded successfully: ${this.embeddingModel}`);
      } catch (error) {
        throw new Error(
          `Failed to load embedding model '${this.embeddingModel}': ${error.message}`
        );
      }
    }
  }

  async exportModelFiles(modelName) {
    try {
      console.log(`Starting export for model: ${modelName}`);

      // Check both Cache API and IndexedDB
      const cacheNames = await caches.keys();
      console.log('Available Cache API caches:', cacheNames);

      // List all IndexedDB databases to find the right one
      const databases = await indexedDB.databases();
      console.log('Available IndexedDB databases:', databases.map(db => db.name));

      // First try Cache API (Transformers.js v3+ uses this)
      const modelCacheName = cacheNames.find(name =>
        name.includes('transformers') ||
        name.includes('huggingface') ||
        name.includes('onnx')
      );

      if (modelCacheName) {
        console.log(`Found model cache: ${modelCacheName}`);
        return await this._exportFromCacheAPI(modelName, modelCacheName);
      }

      // Try different possible database names for IndexedDB (older versions)
      const possibleDbNames = [
        'transformers-cache',
        'transformers_cache',
        'huggingface-assets',
        'transformers.js',
        ...databases.map(db => db.name)
      ];

      let db = null;
      let dbName = null;

      for (const name of possibleDbNames) {
        try {
          db = await new Promise((resolve, reject) => {
            const request = indexedDB.open(name);
            request.onsuccess = () => {
              if (request.result.objectStoreNames.length > 0) {
                resolve(request.result);
              } else {
                request.result.close();
                resolve(null);
              }
            };
            request.onerror = () => resolve(null);
          });
          if (db) {
            dbName = name;
            console.log(`Found database: ${dbName}`);
            break;
          }
        } catch (e) {
          console.log(`Database ${name} not found or empty`);
        }
      }

      if (!db) {
        console.error('No IndexedDB cache found. Available databases:', databases);
        return `Error: No model cache database found. Models may still be downloading or using a different cache mechanism.`;
      }

      console.log(`Object stores in ${dbName}:`, Array.from(db.objectStoreNames));

      const storeName = db.objectStoreNames[0];
      const transaction = db.transaction([storeName], 'readonly');
      const store = transaction.objectStore(storeName);

      const allKeys = await new Promise((resolve, reject) => {
        const request = store.getAllKeys();
        request.onsuccess = () => resolve(request.result);
        request.onerror = () => reject(request.error);
      });

      console.log(`Total cached entries: ${allKeys.length}`);
      console.log(`Sample keys:`, allKeys.slice(0, 5));

      // Filter keys for this model
      const modelKeys = allKeys.filter(key =>
        String(key).includes(`Xenova/${modelName}`) || String(key).includes(modelName)
      );

      console.log(`Found ${modelKeys.length} cached files for ${modelName}`);

      if (modelKeys.length === 0) {
        return `Error: No cached files found for model '${modelName}'. Model may not be cached yet.`;
      }

      // Create directory for model files
      const pyodide = globalThis.pyodide;
      if (!pyodide) {
        return 'Error: Pyodide not available, cannot export to filesystem';
      }

      const modelDir = `/drive/models/Xenova/${modelName}`;
      pyodide.FS.mkdirTree(modelDir);
      console.log(`Created directory: ${modelDir}`);

      let exportedCount = 0;

      // Export each file
      for (const key of modelKeys) {
        const data = await new Promise((resolve, reject) => {
          const request = store.get(key);
          request.onsuccess = () => resolve(request.result);
          request.onerror = () => reject(request.error);
        });

        console.log(`Processing key: ${key}, data type:`, typeof data, data);

        if (data) {
          let arrayBuffer;

          // Handle different data formats
          if (data instanceof Blob) {
            arrayBuffer = await data.arrayBuffer();
          } else if (data.blob instanceof Blob) {
            arrayBuffer = await data.blob.arrayBuffer();
          } else if (data instanceof ArrayBuffer) {
            arrayBuffer = data;
          } else if (data.data instanceof ArrayBuffer) {
            arrayBuffer = data.data;
          } else {
            console.warn(`Unknown data format for key ${key}:`, data);
            continue;
          }

          const uint8Array = new Uint8Array(arrayBuffer);

          // Extract filename from key
          const filename = String(key).split('/').pop();
          const filepath = `${modelDir}/${filename}`;

          pyodide.FS.writeFile(filepath, uint8Array);
          console.log(`Exported: ${filepath} (${uint8Array.length} bytes)`);
          exportedCount++;
        }
      }

      db.close();

      const message = `Successfully exported ${exportedCount} files to ${modelDir}`;
      console.log(message);
      return message;

    } catch (error) {
      const errorMsg = `Export failed: ${error.message}`;
      console.error(errorMsg, error);
      return errorMsg;
    }
  }

  async _exportFromCacheAPI(modelName, cacheName) {
    try {
      const cache = await caches.open(cacheName);
      const requests = await cache.keys();

      console.log(`Total cached requests: ${requests.length}`);

      // Filter for this model
      const modelRequests = requests.filter(req =>
        req.url.includes(`Xenova/${modelName}`) || req.url.includes(modelName)
      );

      console.log(`Found ${modelRequests.length} cached files for ${modelName}`);

      if (modelRequests.length === 0) {
        return `Error: No cached files found for model '${modelName}' in Cache API.`;
      }

      // Access Pyodide filesystem (in web worker context)
      // Try different ways to access pyodide
      let pyodide = null;
      try {
        pyodide = self.pyodide;
      } catch (e) {}

      if (!pyodide) {
        try {
          pyodide = globalThis.pyodide;
        } catch (e) {}
      }

      if (!pyodide) {
        console.error('Pyodide not found in self or globalThis');
        // Return data to Python side instead
        return await this._exportViaPython(modelName, cacheName, modelRequests);
      }

      console.log('Pyodide found, exporting via FS:', pyodide);

      const modelDir = `/drive/models/Xenova/${modelName}`;

      // Create directory using Pyodide FS
      try {
        pyodide.FS.mkdirTree(modelDir);
        console.log(`Created directory: ${modelDir}`);
      } catch (e) {
        console.error('Failed to create directory:', e);
        return `Error: Failed to create directory ${modelDir}: ${e.message}`;
      }

      let exportedCount = 0;

      for (const request of modelRequests) {
        const response = await cache.match(request);
        if (response) {
          const arrayBuffer = await response.arrayBuffer();
          const uint8Array = new Uint8Array(arrayBuffer);

          // Extract filename from URL
          const url = new URL(request.url);
          const pathParts = url.pathname.split('/');
          const filename = pathParts[pathParts.length - 1];

          const filepath = `${modelDir}/${filename}`;
          pyodide.FS.writeFile(filepath, uint8Array);
          console.log(`Exported: ${filepath} (${uint8Array.length} bytes)`);
          exportedCount++;
        }
      }

      const message = `Successfully exported ${exportedCount} files to ${modelDir}`;
      console.log(message);
      return message;

    } catch (error) {
      const errorMsg = `Cache API export failed: ${error.message}`;
      console.error(errorMsg, error);
      return errorMsg;
    }
  }

  async _exportViaPython(modelName, cacheName, modelRequests) {
    // Export by returning data to Python side
    console.log('Exporting via Python side');

    const cache = await caches.open(cacheName);
    const files = [];

    for (const request of modelRequests) {
      const response = await cache.match(request);
      if (response) {
        const arrayBuffer = await response.arrayBuffer();
        const uint8Array = new Uint8Array(arrayBuffer);

        // Extract filename from URL
        const url = new URL(request.url);
        const pathParts = url.pathname.split('/');
        const filename = pathParts[pathParts.length - 1];

        files.push({
          filename: filename,
          data: uint8Array
        });
      }
    }

    return {
      success: true,
      model: modelName,
      files: files
    };
  }

  async generate(prompt) {
    if (!this.generator) {
      throw new Error("Generator model not initialized. Call init() first.");
    }

    try {
      const output = await this.generator(prompt, {
        max_new_tokens: this.max_new_tokens,
        temperature: this.temperature,
        top_k: this.top_k,
        do_sample: this.do_sample
      });

      return output[0].generated_text;
    } catch (error) {
      throw new Error(
        `Text generation failed: ${error.message}. ` +
        `This could be due to: ` +
        `(1) Model still downloading - wait and retry, ` +
        `(2) Memory constraints in browser, ` +
        `(3) ONNX Runtime error - try a different model like 'gpt2', ` +
        `(4) Model incompatibility with browser WASM environment.`
      );
    }
  }

  async embed(text) {
    if (!this.embedder) {
      throw new Error("No embedding model configured");
    }

    const result = await this.embedder(text);
    return Array.from(result.data);
  }
}

// --- Factory ---

function createLLM(config) {
  return new BrowserLLM(config);
}

// --- Similarity helpers ---

function cosineSimilarity(a, b) {
  let dot = 0.0, normA = 0.0, normB = 0.0;

  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

function cosineSimilarityBatch(queryVec, docVecs) {
  return docVecs.map(doc => cosineSimilarity(queryVec, doc));
}

// --- expose globally (critical for Python) ---

globalThis.createLLM = createLLM;
globalThis.cosineSimilarity = cosineSimilarity;
globalThis.cosineSimilarityBatch = cosineSimilarityBatch;

undefined;

